### PR TITLE
add `conflictsWith` to provider schema

### PR DESCRIPTION
**Pull request by BBBmau**
### Description

This resolves the issue that comes up when wanting to use 2 k8s configurations 

Relates to #1141, updates the original PR

### Acceptance tests
- [ ] Have you added an acceptance test for the functionality being added?
- [ ] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/CHANGELOG.md):

```release-note
...
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

File: kubernetes/provider.go [hunk 1]
```diff
@@ -40,46 +40,58 @@ func Provider() *schema.Provider {
...
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("KUBE_CLIENT_KEY_DATA", ""),
-				Description: "PEM-encoded client certificate key for TLS authentication.",
+				Type:          schema.TypeString,
+				Optional:      true,
+				DefaultFunc:   schema.EnvDefaultFunc("KUBE_CLIENT_KEY_DATA", nil),
+				Description:   "PEM-encoded client certificate key for TLS authentication.",
+				ConflictsWith: []string{"config_path", "config_paths", "username", "password", "exec", "insecure"},
+				RequiredWith:  []string{"client_certificate", "cluster_ca_certificate", "host"},
```
Comment (alexsomesan) [inline kubernetes/provider.go:86]: `cluster_ca_certificate` isn't related to the client certificate credentials. It's pretty much independent and shouldn't depend on be required by any of the attributes, AFAIK.

File: kubernetes/provider.go [hunk 2]
```diff
@@ -40,46 +40,57 @@ func Provider() *schema.Provider {
...
 			"client_key": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("KUBE_CLIENT_KEY_DATA", ""),
-				Description: "PEM-encoded client certificate key for TLS authentication.",
+				Type:          schema.TypeString,
+				Optional:      true,
+				DefaultFunc:   schema.EnvDefaultFunc("KUBE_CLIENT_KEY_DATA", nil),
+				Description:   "PEM-encoded client certificate key for TLS authentication.",
+				ConflictsWith: []string{"config_path", "config_paths", "username", "password", "exec", "insecure"},
```
Comment (alexsomesan) [inline kubernetes/provider.go:85]: I'm not 100% sure that `insecure` is a conflict in this case. TLS certs can be client-side only in which case they aren't involved in any server auth. The `insecure` flag would still be needed if the API doesn't have a trusted server-side cert.

File: kubernetes/provider.go [hunk 3]
```diff
@@ -40,46 +40,57 @@ func Provider() *schema.Provider {
...
 			"client_certificate": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("KUBE_CLIENT_CERT_DATA", ""),
-				Description: "PEM-encoded client certificate for TLS authentication.",
+				Type:          schema.TypeString,
+				Optional:      true,
+				DefaultFunc:   schema.EnvDefaultFunc("KUBE_CLIENT_CERT_DATA", nil),
+				Description:   "PEM-encoded client certificate for TLS authentication.",
+				ConflictsWith: []string{"config_path", "config_paths", "username", "password", "insecure"},
```
Comment (alexsomesan) [inline kubernetes/provider.go:77]: I'm not 100% sure that `insecure` is a conflict in this case. TLS certs can be client-side only in which case they aren't involved in any server auth. The `insecure` flag would still be needed if the API doesn't have a trusted server-side cert.

File: kubernetes/provider.go [hunk 4]
```diff
@@ -40,46 +40,57 @@ func Provider() *schema.Provider {
...
 			"insecure": {
-				Type:        schema.TypeBool,
-				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("KUBE_INSECURE", false),
-				Description: "Whether server should be accessed without verifying the TLS certificate.",
+				Type:          schema.TypeBool,
+				Optional:      true,
+				DefaultFunc:   schema.EnvDefaultFunc("KUBE_INSECURE", nil),
+				Description:   "Whether server should be accessed without verifying the TLS certificate.",
+				ConflictsWith: []string{"client_key", "client_certificate", "exec"},
```
Comment (alexsomesan) [inline kubernetes/provider.go:70]: See comments on client_cert

File: kubernetes/provider.go [hunk 5]
```diff
@@ -40,46 +40,57 @@ func Provider() *schema.Provider {
...
-				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("KUBE_CLUSTER_CA_CERT_DATA", ""),
-				Description: "PEM-encoded root certificates bundle for TLS authentication.",
+				Type:          schema.TypeString,
+				Optional:      true,
+				DefaultFunc:   schema.EnvDefaultFunc("KUBE_CLUSTER_CA_CERT_DATA", nil),
+				Description:   "PEM-encoded root certificates bundle for TLS authentication.",
+				ConflictsWith: []string{"config_path", "config_paths", "insecure"},
 			},
 			"config_paths": {
```
Comment (alexsomesan) [inline kubernetes/provider.go:95]: There are a couple of attributes above that include this in their `conflicts_with` lists. I wonder if these should be mutually exclusive and they would also need to be mentioned here... I'm not 100% sure.

Comment (BBBmau) [reply kubernetes/provider.go:95]: The way it's designed is to have `ConflictsWith` / `RequiredWith` to mutually exclusive between attributes. I was able to find an example from another provider that shows a good example of [this](https://github.com/hashicorp/terraform-provider-nomad/blob/77c12b04db15efd9f7bc84d13f229e9bba861c1c/nomad/provider.go#L47)

File: kubernetes/provider.go (hunk 4 above)
Comment (alexsomesan) [reply kubernetes/provider.go:70]: @BBBmau I think this needs changing. `client_certificate`, `client_key` and `exec` are not incompatible with `insecure`
IMO, only `cluster_ca_certificate` is a conflict.

File: kubernetes/provider.go [hunk 6]
```diff
@@ -40,46 +40,57 @@ func Provider() *schema.Provider {
...
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("KUBE_CLIENT_KEY_DATA", ""),
-				Description: "PEM-encoded client certificate key for TLS authentication.",
+				Type:          schema.TypeString,
+				Optional:      true,
+				DefaultFunc:   schema.EnvDefaultFunc("KUBE_CLIENT_KEY_DATA", nil),
+				Description:   "PEM-encoded client certificate key for TLS authentication.",
+				ConflictsWith: []string{"config_path", "config_paths", "username", "password", "exec"},
+				RequiredWith:  []string{"client_certificate", "host"},
```
Comment (alexsomesan) [inline kubernetes/provider.go:86]: same here, "host" isn't a requirement for the same reason as the cert.

File: kubernetes/provider.go [hunk 7]
```diff
@@ -40,46 +40,57 @@ func Provider() *schema.Provider {
 	p := &schema.Provider{
 		Schema: map[string]*schema.Schema{
 			"host": {
```
Comment (jrhouston) [inline kubernetes/provider.go:42]: When i tried this out I was still able to set both `host` and `config_path`.

I think you need to set ConflictsWith for both attributes, i.e you need to add `host` to the ConflictsWith for `config_path`. You'll need to this for the other attributes you've added this too as well.

**Review by jrhouston — CHANGES_REQUESTED** [review]

**Review by jrhouston — APPROVED** [review]

**Comment by aquarhead** [comment]
While I totally agree this change is technically correct, the old behaviour has a useful side effect.

What I've been doing for years is to only set `host` in the provider configuration:

```
provider "kubernetes" {
  host = "..."
}
```

The actual auth will be configured through env vars (e.g. I use `KUBE_CONFIG_PATH` and `KUBE_CTX` env var). And if the kube cluster the auth/context refers to is different than the `host`, it will error out thus preventing deploying things in the wrong kube cluster by accident. (The error is actually something related to certificate, but once you know what it means it serves the purpose pretty well)

This is useful in the sense that it allows Terraform code to enforce the exact cluster a config is supposed to be used, while at the same time allowing auth information to be configured through potentially different manners, not just between developers, but between local and CI as well (which iirc was certainly a case at my previous job). (And yes it did happen that we used a wrong config in the wrong cluster, hence I looked around for way to prevent it)

I'm a bit upset that this is not released as a major version even though I'll be the first to admit I was abusing the behaviour, still, I think the ability to enforce the correct cluster without hardcoding the auth has some merit.

I would like to propose an alternative solution, instead of using `conflictsWith` directly, could we try to merge all the different config options, and error out if anything mismatch? In my use case this will actually give useful error message like "`host` has conflict with context" but only if the wrong context is loaded.

---

Before you ask, I run EKS and I'm aware of the `exec` config method. But in general I'd like Terraform code itself to not enforce how authentication to be done. Also, if the terraform provider is fully configured, there could be a situation where the Terraform operates on a different cluster than `kubectl` is, that can be quite confusing - it happens a lot that you deploy through tf and then check things through `kubectl`. I think it's nice to configure both at the same time, e.g. by pointing to a context.

---

I want to stress again I totally see where this change is coming from, but I want to explain my (ab)use case and I do believe it will be helfpul especially for people operating many kube clusters, to be able to enforce the correct cluster in code, while keeping the auth flexible. (Maybe, we could introduce a `host_match` or probably `certificate_match` would be more correct? :thinking:)

I will totally understand if it doesn't happen, but I do wish that this change be reverted in this major version and explore some other options instead, or introduce only in the next major version. We do not use Terraform lockfiles because of a whole slew of other issues, and in general we would like to use latest providers (locked to major version).

**Comment by Pell17** [comment]
Breaking change for us. Would have been nice if this could have initially been a Warning instead of Error, giving some time to "fix" our configuration. We usually auto-upgrade providers to the newest minor version, not expecting things to break, as required by SemVer.

**Comment by emoshaya** [comment]
breaking change for us too :(

**Comment by Nuru** [comment]
I have a few problems with this.

As others have mentioned, this is a breaking change in a minor version release, which violates SemVer, and has caused a lot of disruption with no notice.

The documentation has not been updated to indicate what conflicts with what. For example, does `username` conflict with `client_key`? 

This change removes our ability to override environment variables. In [#1175](https://github.com/hashicorp/terraform-provider-kubernetes/issues/1175#issuecomment-786121967), @dak1n1 explained that v2 of this provider consciously avoided taking configuration from the `KUBECONFIG` environment variable to avoid having it direct Terraform to the wrong cluster accidentally. (This provider uses `KUBE_CONFIG_PATH` and `KUBE_CONFIG_PATHS` instead.) In our modules, we want to further ensure the environment variables are not used, and we have been doing that by setting `config_path = ""`. This change breaks that.

After we fix the breakage by removing `config_path`, now when both `host` and `KUBE_CONFIG_PATH` are set, we get the confusing error message:

```
│ Error: Conflicting configuration arguments
│ 
│   with provider["registry.terraform.io/hashicorp/kubernetes"],
│   on providers.tf line 14, in provider "kubernetes":
│  14:   host                   = local.eks_cluster_endpoint
│ 
│ "host": conflicts with config_path
╵
```

(Confusing because `config_path` is not present in the providers block. A more helpful error would be "host conflicts with KUBE_CONFIG_PATH".)


### Suggestion

I believe the better path forward is to arrange the configuration options in a hierarchy, rather than cause them to conflict. Ideally the provider would use configuration provided directly in the provider block in preference to environment variables, and not complain (or perhaps issue a warning) if the environment variables provide a different configuration than what is specified in the provider block. This would allow the provider to support `KUBECONFIG` as has been requested more than once (#1973, #1175, and perhaps more) while avoiding the problem that environment variables override specifically configured authentication. But perhaps that is best deferred until v3 of this provider.

Until v3 is ready, please revert this change and discuss backwards compatible improvements instead.

**Comment by jrhouston** [comment]
Thank you for your thoughtful responses @aquarhead @Pell17 @Nuru – we are going to revert this change for now. We will put out a hotfix release once the above PR is merged.

**Comment by ingnil** [comment]
This change also does not care about the value of the parameters. I just ran into a conflict between `cluster_ca_certificate` and `insecure` despite the latter being set to `false`.

**Pull request merged**